### PR TITLE
mcdropout predict sample moved to uanets

### PR DESCRIPTION
**Pull request by hstojic**
implementing interface methods

something to think about is what kind of unit tests to have, any suggestions?
one thing is that there is not classification tasks here and probably there should be

### CODE REVIEW

File: tests/uanets/models/test_mc_dropout.py [hunk 1]
```diff
@@ -142,3 +148,65 @@ def test_dropout_network_accepts_scalars() -> None:
     test_points = tf.linspace(-1, 1, 100)
 
     assert model(test_points).shape == (100, 1)
+
+
+@pytest.mark.parametrize(
+    "input_shape, output_shape",
+    [
+        [[1, 1], [1, 1]],
+        [[10, 5], [10, 3]],
```
Comment (hstojic) [inline tests/uanets/models/test_mc_dropout.py:157]: I'm perhaps not general enough here, this is something that is geared toward handling trieste use cases?

File: uanets/models/mc_dropout.py [hunk 2]
```diff
@@ -111,3 +115,79 @@ def call(self, inputs: tf.Tensor) -> tf.Tensor:
         output = self.output_layer(hidden_output)
 
         return output
+
+    @inherit_check_shapes
+    def predict_mean_and_var(
+        self, x: tf.Tensor, num_samples: Optional[int] = None, seed: Optional[int] = None
```
Comment (hstojic) [inline uanets/models/mc_dropout.py:121]: typing question here - for this model `num_samples` is actually required, but if I remove `Optional[]` I get inconsistency with parent error, if I leave it like this I get typing errors with `num_samples` below - not sure what to do...

Comment (willcowley) [reply uanets/models/mc_dropout.py:121]: I'd probably just have either an `assert num_samples is not None` or
```
if num_samples is None:
    raise TypeError("num_samples must be set for this method")
```
somewhere in the implementation (before `num_samples` is used...you may also want to have a similar check for `num_samples > 0`).

Alternatively, we could change the method signature in the abc to be `num_samples: int = 1_000` (or whatever default we think is sensible)

File: tests/util/misc.py [hunk 3]
```diff
@@ -99,3 +99,18 @@ def inputs_outputs_spec(
     inputs = tf.TensorSpec(tf.TensorShape([0]) + inputs_shape, dtype)
     outputs = tf.TensorSpec(tf.TensorShape([0]) + outputs_shape, dtype)
     return inputs, outputs
+
+
+def random_inputs_outputs(
+    inputs_shape: ShapeLike, outputs_shape: ShapeLike, dtype: tf.DType = tf.float64
```
Comment (willcowley) [inline tests/util/misc.py:105]: Are we expecting `inputs_shape` and `outputs_shape` to be the same apart from the trailing dimension (i.e. `inputs_shape[:-1] == outputs_shape[:-1]` if so do we want to document/assert this?

File: uanets/models/mc_dropout.py [hunk 4]
```diff
@@ -111,3 +115,79 @@ def call(self, inputs: tf.Tensor) -> tf.Tensor:
...
+        stochastic_passes = tf.stack(
+            [self.__call__(x, training=True) for _ in range(num_samples)], axis=0
+        )
+        predicted_means = tf.math.reduce_mean(stochastic_passes, axis=0)
+
+        predicted_vars = tf.subtract(
+            tf.divide(
+                tf.reduce_sum(tf.math.multiply(stochastic_passes, stochastic_passes), axis=0),
+                num_samples,
+            ),
```
Comment (willcowley) [inline uanets/models/mc_dropout.py:163]:
```suggestion
            tf.reduce_mean(tf.math.square(stochastic_passes), axis=0),
```

File: tests/uanets/models/test_mc_dropout.py [hunk 5]
```diff
@@ -142,3 +148,65 @@ def test_dropout_network_accepts_scalars() -> None:
...
+@pytest.mark.parametrize("num_samples", [100, 500, 1000])
+def test_dropout_network_sample_equal_predict(rate: float, num_samples) -> None:
+    input_shape, output_shape = [100, 5], [100, 3]
+    inputs, outputs = random_inputs_outputs(input_shape, output_shape)
+
+    model = mc_dropout_model(inputs, outputs, rate)
+
+    samples = model.sample(inputs, num_samples)
+    sample_mean = tf.reduce_mean(samples, axis=0)
+    sample_variance = tf.reduce_mean((samples - sample_mean) ** 2, axis=0)
```
Comment (willcowley) [inline tests/uanets/models/test_mc_dropout.py:205]:
```suggestion
    sample_variance = tf.math.reduce_variance(samples, axis=0)
```
Actually, could we just use `reduce_variance` in the actual `predict_mean_and_var` implementation?

File: uanets/models/mc_dropout.py [hunk 6]
```diff
@@ -111,3 +115,79 @@ def call(self, inputs: tf.Tensor) -> tf.Tensor:
...
+            The number of stochastic passes based on which mean and variance will be computed.
+        :param seed:
+            The seed to produce deterministic results - unused here.
+        :return: The mean and variance of the independent marginal distributions at each point in
+            ``x``.
+        """
+        tf.debugging.assert_greater_equal(num_samples, 1)
+        stochastic_passes = tf.stack(
+            [self.__call__(x, training=True) for _ in range(num_samples)], axis=0
+        )
```
Comment (willcowley) [inline uanets/models/mc_dropout.py:156]: Could we just call `self.sample` here?

File: uanets/models/mc_dropout.py (hunk 2 above)
Comment (hstojic) [reply uanets/models/mc_dropout.py:121]: I'll opt for first one I think

File: tests/util/misc.py (hunk 3 above)
Comment (hstojic) [reply tests/util/misc.py:105]: yes, we are I guess, I'll document it and assert it

File: uanets/models/mc_dropout.py (hunk 4 above)
Comment (hstojic) [reply uanets/models/mc_dropout.py:163]: indeed :)

File: tests/uanets/models/test_mc_dropout.py (hunk 5 above)
Comment (hstojic) [reply tests/uanets/models/test_mc_dropout.py:205]: that looks correct

File: uanets/models/mc_dropout.py (hunk 6 above)
Comment (hstojic) [reply uanets/models/mc_dropout.py:156]: true, we can do that

**Pull request merged**